### ISSUE TITLE
demo_database.sql - ERROR 1452 (23000) at line 5

### ISSUE THREAD

**Issue by PierrickV** [rapid7/hackazon#13]
Hi,

When I try to import demo_database.sql I have the following error :

> hackazon_db_1   | ERROR 1452 (23000) at line 5: Cannot add or update a child row: a foreign key constraint fails (`hackazon`.`tbl_enquiries`, CONSTRAINT `tbl_enquiries_ibfk_1` FOREIGN KEY (`created_by`) REFERENCES `tbl_users` (`id`) ON DELETE CASCADE ON UPDATE CASCADE) 

Full Docker logs:  [db_error.txt](https://github.com/rapid7/hackazon/files/604815/db_error.txt)

Can you help me with this ?

The DB is working on this Docker image : [https://github.com/PierrickV/hackazon](https://github.com/PierrickV/hackazon)

Thanks

**Comment by PierrickV** [comment]
Same error with manual importation:
```bash
root@248c3ddc86e0:/tmp# mysql --user=hackazon --password=hackazon --database=hackazon --host=hackazon_db < db.sql 
root@248c3ddc86e0:/tmp# wget https://raw.githubusercontent.com/rapid7/hackazon/master/database/demo_database.sql
converted 'https://raw.githubusercontent.com/rapid7/hackazon/master/database/demo_database.sql' (ANSI_X3.4-1968) -> 'https://raw.githubusercontent.com/rapid7/hackazon/master/database/demo_database.sql' (UTF-8)
--2016-11-23 18:46:03--  https://raw.githubusercontent.com/rapid7/hackazon/master/database/demo_database.sql
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 151.101.16.133
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|151.101.16.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 271140 (265K) [text/plain]
Saving to: 'demo_database.sql'

demo_database.sql                                         100%[======================================================================================================================================>] 264.79K  --.-KB/s   in 0.1s   

2016-11-23 18:46:03 (1.85 MB/s) - 'demo_database.sql' saved [271140/271140]

root@248c3ddc86e0:/tmp# mysql --user=hackazon --password=hackazon --database=hackazon --host=hackazon_db < demo_database.sql 
ERROR 1452 (23000) at line 5: Cannot add or update a child row: a foreign key constraint fails (`hackazon`.`tbl_enquiries`, CONSTRAINT `tbl_enquiries_ibfk_1` FOREIGN KEY (`created_by`) REFERENCES `tbl_users` (`id`) ON DELETE CASCADE ON UPDATE CASCADE)
root@248c3ddc86e0:/tmp# 
```

**mysql:5.5**

**Comment by gmaran23** [comment]
Hello

I have had a similar foreign key contraint error. When I ran the database script, the script executed and created some tables, however failed half way. When I tried to re-run the script, then I entounceted the foreign key contraint error. I deleted the hackazon database and then after fixing the script I re-created the database and ran the script again.

Fixing the script involved setting the mysql allow invalid date flag. May that will help?

my experience here - https://renouncedthoughts.wordpress.com/2017/02/21/getting-hackazon-up-and-running-on-wamp-on-nov-2016/

**Comment by PierrickV** [comment]
Hi,

Thank you for your help and for sharing your experience.

I tried to add the line :
```bash
SET @OLD_SQL_MODE=@@SQL_MODE, SQL_MODE='ALLOW_INVALID_DATES';
```
To db.sql and demo.sql but I have the same error.

```bash
root@8b14344805fe:/db# mysql --user=hackazon --password=hackazon --database=hackazon --host=hackazon_db < db.sql 
root@8b14344805fe:/db# mysql --user=hackazon --password=hackazon --database=hackazon --host=hackazon_db < demo_database.sql 
ERROR 1452 (23000) at line 6: Cannot add or update a child row: a foreign key constraint fails (`hackazon`.`tbl_enquiries`, CONSTRAINT `tbl_enquiries_ibfk_1` FOREIGN KEY (`created_by`) REFERENCES `tbl_users` (`id`) ON DELETE CASCADE ON UPDATE CASCADE)
```

I also tried to find some help on rapid7's forum but no answer yet :
https://community.rapid7.com/message/26331

**Comment by gmaran23** [comment]
Hi,

"ERROR 1452 (23000) at line 6: Cannot add or update a child row: a foreign key constraint fails (`hackazon`.`tbl_enquiries`, CONSTRAINT `tbl_enquiries_ibfk_1` FOREIGN KEY (`created_by`) REFERENCES `tbl_users` (`id`) ON DELETE CASCADE ON UPDATE CASCADE)
"

It seems like some tables/columns required for the foreign key constraint are missing. Could you try to delete the hackazon databse and then re-run the .sql file.

I have attached the db file that i used. I don't see demo_database.sql in my hackazon folder. Where did you pick that file from?

[db - Working.sql.txt](https://github.com/rapid7/hackazon/files/816376/db.-.Working.sql.txt)

**Comment by PierrickV** [comment]
When your installation is done, are there any items in the shop ?

I tried to import demo_database because mine was empty. The file is here : 
https://github.com/rapid7/hackazon/tree/master/database

**Comment by gmaran23** [comment]
Yes I did have some items in the store, some FAQs and other things.

ok. now I see where the demo database comes from. I never had to use it though. I followed the instruction guide from the rapid  community (as mentioned in my blog) and edited a few things in the db.sql file (as mentioned in blog) and fixed the issues one by one.

I only see a bunch of INSERT statements in the https://github.com/rapid7/hackazon/blob/master/database/demo_database.sql 

Looking at the foreign key contraint error message it appears that there were enquiries made by a user, and that user does not exist in the users table.

It is just a quick guess. I haven't looked around all the insert statements. But looking at the foreign key contraint error message the below may work.

Cut line # 18 and 19 from https://github.com/rapid7/hackazon/blob/master/database/demo_database.sql 

and try to paste it at the top before executing the first line INSERT INTO `tbl_cart` VALUES ....

and see what happens.

Anyways, I did not have to use the demo_database.sql in my installation. And yet i still had products. 

Which stage of installation are you in?

Maran
